### PR TITLE
Improvements and upgrades

### LINKED ISSUE #247: Error message should be based on exception, not hardcoded
https://github.com/pennlabs/common-funding-application/pull/245/files#diff-b5099d1f0106025f31781ccbe0a59d13R151

### LINKED ISSUE #253: Ability to change password, and reset password if forgotten

### LINKED ISSUE #262: Event time/date selector does not work well at all
Selector does not work very frequently, a pain to use actually.

### LINKED ISSUE #272: Remove use of __unicode__ and unicode

### PR DESCRIPTION
**Pull request by ezwang**
- More specific error handling when creating/editing applications
  - Doesn't 500 when submitting invalid input anymore
  - Shows form specific errors for fields at top
- Partially restores form if errors are encountered (only top fields)
- Change login page to homepage
- Upgrade various javascript libraries
  - jQuery, underscore, pickadate, timepicker
- Upgrade to Bootstrap 4
- Merge and cleanup various templates
  - events + events_old merged
  - various registration templates added (password reset, password change)
- Implement password changing and reset functionality
- Added confirmation before deleting applications

Fixes #272, fixes #262, fixes #253, fixes #247.

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -60,12 +60,12 @@ class CFAUser(models.Model):
...
 
-    def __unicode__(self):
+    def __str__(self):
         if self.is_funder:
             # some funders might not have funder_names
-            return unicode(self.funder_name) or unicode(self.user)
+            return str(self.funder_name) or str(self.user)
         else:
-            return unicode(self.user)
+            return str(self.user)
```
Comment (adelq) [inline app/models.py]: Does this work in Python 2 with unicode/special characters? It doesn't need to, I'm just wondering.

File: app/static/css/form.css [hunk 2]
```diff
@@ -58,29 +93,21 @@ fieldset {
...
   cursor: pointer;
   margin-top: 10px;
 }
 .section-toggle:after {
-  content: "collapse";
+  content: "\f068";
 }
 .section-toggle.collapsed:after {
-  content: "expand";
+  content: "\f067";
```
Comment (adelq) [inline app/static/css/form.css]: These characters are not rendering for me, not sure why. Other FontAwesome things like the footer work fine.

File: app/static/css/index.css [hunk 3]
```diff
@@ -1,20 +1,39 @@
...
+    color: #495057;
+    background-color: #fff;
+    background-clip: padding-box;
+    border: 1px solid #ced4da;
+    border-radius: .25rem;
+    transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
+}
+#login input:focus {
+    outline: none;
+    border-color: red;
```
Comment (adelq) [inline app/static/css/index.css]: Why red? Makes me feel like there's an error.

File: app/models.py (hunk 1 above)
Comment (ezwang) [reply app/models.py]: This doesn't work with Python 2, unfortunately. I could keep in both __unicode__ and __str__, but this seemed repetitive to me.

File: app/static/css/index.css (hunk 3 above)
Comment (ezwang) [reply app/static/css/index.css]: The original application before the changes also had a red border when a text box was selected, so I thought it was part of the style. Turns out it only turns red when the text box is empty. Nice catch, will fix.

**Pull request merged**